Refactor out headeronly ArrayRef

janeyx99 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165953
* #165153
* #165152
* __->__ #164991

Differential Revision: [D85091961](https://our.internmc.facebook.com/intern/diff/D85091961)

janeyx99 committed:
Refactor out headeronly ArrayRef

[ghstack-poisoned]

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"




[ghstack-poisoned]

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"




[ghstack-poisoned]

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"




[ghstack-poisoned]

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"




[ghstack-poisoned]

swolchok commented on c10/util/ArrayRef.h:
```diff
@@ -40,200 +41,105 @@ namespace c10 {
+/// into HeaderOnlyArrayRef. As adding `virtual` would change the performance of
+/// the underlying constexpr calls, we rely on apparent-type dispatch for
+/// inheritance. This should be fine because their memory format is the same,
+/// and it is never incorrect for ArrayRef to call HeaderOnlyArrayRef methods.
```
I think we should probably add something like "However, you should prefer to use ArrayRef when possible, because its use of TORCH_CHECK will lead to better user-facing error messages."

swolchok commented on torch/headeronly/util/HeaderOnlyArrayRef.h:
```diff
@@ -0,0 +1,246 @@
+/// 1. ArrayRef has an extra convenience constructor for SmallVector.
+/// 2. ArrayRef uses TORCH_CHECK. HeaderOnlyArrayRef uses header-only
+/// STD_TORCH_CHECK,
+///    which will output a std::runtime_error vs a c10::Error.
```
and here, something like "Consequently, you should use ArrayRef when possible and HeaderOnlyArrayRef only when necessary to support headeronly code."

swolchok commented on c10/util/ArrayRef.h:
```diff
@@ -40,200 +41,105 @@ namespace c10 {
-      : Data(data), Length(length) {
-    debugCheckNullptrInvariant();
-  }
+  // Inherits all constructors from HeaderOnlyArrayRef<T>
```
I don't love comments that explain C++ language features

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"




[ghstack-poisoned]

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"




[ghstack-poisoned]

janeyx99 committed:
Update on "Refactor out headeronly ArrayRef"


Differential Revision: [D85091961](https://our.internmc.facebook.com/intern/diff/D85091961)

[ghstack-poisoned]
